File: prd-api/src/PrdAgent.Core/Models/TutorialEmail.cs
namespace PrdAgent.Core.Models;

/// <summary>
/// 教程邮件序列定义（如 "新用户引导"、"功能更新通知"）
/// </summary>
public class TutorialEmailSequence
{
    public string Id { get; set; } = Guid.NewGuid().ToString("N");

    /// <summary>序列唯一标识（如 onboarding、feature-update-v2）</summary>
    public string SequenceKey { get; set; } = string.Empty;

    /// <summary>序列名称（管理后台展示用）</summary>
    public string Name { get; set; } = string.Empty;

    /// <summary>序列描述</summary>
    public string? Description { get; set; }

    /// <summary>触发类型：registration=注册后自动、manual=手动、feature-release=版本发布</summary>
    public string TriggerType { get; set; } = "manual";

    /// <summary>序列步骤（按 dayOffset 排序）</summary>
    public List<TutorialEmailStep> Steps { get; set; } = new();

    /// <summary>是否启用</summary>
    public bool IsActive { get; set; } = true;

    public string? CreatedBy { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

/// <summary>
/// 序列中的单个步骤
/// </summary>
public class TutorialEmailStep
{
    /// <summary>第几天发送（相对于 enrollment 开始日期，0=当天）</summary>
    public int DayOffset { get; set; }

    /// <summary>邮件标题</summary>
    public string Subject { get; set; } = string.Empty;

    /// <summary>邮件模板 ID</summary>
    public string TemplateId { get; set; } = string.Empty;

    /// <summary>跳过条件描述（预留，可扩展为表达式引擎）</summary>
    public string? SkipCondition { get; set; }
}

/// <summary>
/// 邮件模板（HTML 内容 + 变量定义）
/// </summary>
public class TutorialEmailTemplate
{
    public string Id { get; set; } = Guid.NewGuid().ToString("N");

    /// <summary>模板名称</summary>
    public string Name { get; set; } = string.Empty;

    /// <summary>HTML 内容（支持 {{变量}} 占位符）</summary>
    public string HtmlContent { get; set; } = string.Empty;

    /// <summary>模板变量列表（如 userName, productName）</summary>
    public List<string> Variables { get; set; } = new();

    /// <summary>关联的截图资源 ID 列表</summary>
    public List<string> AssetIds { get; set; } = new();

    /// <summary>预览用缩略图 URL</summary>
    public string? ThumbnailUrl { get; set; }

    public string? CreatedBy { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

/// <summary>
/// 截图/图片素材
/// </summary>
public class TutorialEmailAsset
{
    public string Id { get; set; } = Guid.NewGuid().ToString("N");

    /// <summary>文件名</summary>
    public string FileName { get; set; } = string.Empty;

    /// <summary>CDN/OSS 公网 URL</summary>
    public string FileUrl { get; set; } = string.Empty;

    /// <summary>标签（便于按功能/版本归类）</summary>
    public List<string> Tags { get; set; } = new();

    /// <summary>文件大小（字节）</summary>
    public long FileSize { get; set; }

    /// <summary>MIME 类型</summary>
    public string? ContentType { get; set; }

    public string? UploadedBy { get; set; }
    public DateTime UploadedAt { get; set; } = DateTime.UtcNow;
}

/// <summary>
/// 用户邮件序列订阅记录
/// </summary>
public class TutorialEmailEnrollment
{
    public string Id { get; set; } = Guid.NewGuid().ToString("N");

    /// <summary>用户 ID</summary>
    public string UserId { get; set; } = string.Empty;

    /// <summary>目标邮箱</summary>
    public string Email { get; set; } = string.Empty;

    /// <summary>序列 Key</summary>
    public string SequenceKey { get; set; } = string.Empty;

    /// <summary>当前已完成步骤索引（-1=尚未开始）</summary>
    public int CurrentStepIndex { get; set; } = -1;

    /// <summary>状态：active=进行中、completed=已完成、unsubscribed=已退订</summary>
    public string Status { get; set; } = "active";

    /// <summary>下次发送时间</summary>
    public DateTime? NextSendAt { get; set; }

    /// <summary>订阅开始时间</summary>
    public DateTime EnrolledAt { get; set; } = DateTime.UtcNow;

    /// <summary>已发送历史</summary>
    public List<TutorialEmailSentRecord> SentHistory { get; set; } = new();

    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

/// <summary>
/// 已发送邮件记录
/// </summary>
public class TutorialEmailSentRecord
{
    /// <summary>步骤索引</summary>
    public int StepIndex { get; set; }

    /// <summary>发送时间</summary>
    public DateTime SentAt { get; set; } = DateTime.UtcNow;

    /// <summary>是否发送成功</summary>
    public bool Success { get; set; }

    /// <summary>失败原因</summary>
    public string? ErrorMessage { get; set; }
}

File: prd-admin/src/services/real/tutorialEmail.ts
import { apiRequest } from '@/services/real/apiClient';
import { api } from '@/services/api';
import type { ApiResponse } from '@/types/api';

// ========== Types ==========

export type TutorialEmailStep = {
  dayOffset: number;
  subject: string;
  templateId: string;
  skipCondition?: string | null;
};

export type TutorialEmailSequence = {
  id: string;
  sequenceKey: string;
  name: string;
  description?: string | null;
  triggerType: string;
  steps: TutorialEmailStep[];
  isActive: boolean;
  createdBy?: string | null;
  createdAt: string;
  updatedAt: string;
};

export type TutorialEmailTemplate = {
  id: string;
  name: string;
  htmlContent: string;
  variables: string[];
  assetIds: string[];
  thumbnailUrl?: string | null;
  createdBy?: string | null;
  createdAt: string;
  updatedAt: string;
};

export type TutorialEmailAsset = {
  id: string;
  fileName: string;
  fileUrl: string;
  tags: string[];
  fileSize: number;
  contentType?: string | null;
  uploadedBy?: string | null;
  uploadedAt: string;
};

export type TutorialEmailSentRecord = {
  stepIndex: number;
  sentAt: string;
  success: boolean;
  errorMessage?: string | null;
};

export type TutorialEmailEnrollment = {
  id: string;
  userId: string;
  email: string;
  sequenceKey: string;
  currentStepIndex: number;
  status: string;
  nextSendAt?: string | null;
  enrolledAt: string;
  sentHistory: TutorialEmailSentRecord[];
  updatedAt: string;
};

// ========== Sequences ==========

export async function listTutorialEmailSequences(): Promise<ApiResponse<{ items: TutorialEmailSequence[] }>> {
  return apiRequest(api.tutorialEmail.sequences.list(), { method: 'GET' });
}

export async function getTutorialEmailSequence(id: string): Promise<ApiResponse<TutorialEmailSequence>> {
  return apiRequest(api.tutorialEmail.sequences.byId(id), { method: 'GET' });
}

export async function createTutorialEmailSequence(data: {
  sequenceKey: string;
  name: string;
  description?: string;
  triggerType?: string;
  steps?: TutorialEmailStep[];
  isActive?: boolean;
}): Promise<ApiResponse<TutorialEmailSequence>> {
  return apiRequest(api.tutorialEmail.sequences.list(), { method: 'POST', body: data });
}

export async function updateTutorialEmailSequence(id: string, data: {
  name?: string;
  description?: string;
  triggerType?: string;
  steps?: TutorialEmailStep[];
  isActive?: boolean;
}): Promise<ApiResponse<TutorialEmailSequence>> {
  return apiRequest(api.tutorialEmail.sequences.byId(id), { method: 'PUT', body: data });
}

export async function deleteTutorialEmailSequence(id: string): Promise<ApiResponse<{ deleted: boolean }>> {
  return apiRequest(api.tutorialEmail.sequences.byId(id), { method: 'DELETE' });
}

// ========== Templates ==========

export async function listTutorialEmailTemplates(): Promise<ApiResponse<{ items: TutorialEmailTemplate[] }>> {
  return apiRequest(api.tutorialEmail.templates.list(), { method: 'GET' });
}

export async function getTutorialEmailTemplate(id: string): Promise<ApiResponse<TutorialEmailTemplate>> {
  return apiRequest(api.tutorialEmail.templates.byId(id), { method: 'GET' });
}

export async function createTutorialEmailTemplate(data: {
  name: string;
  htmlContent: string;
  variables?: string[];
  assetIds?: string[];
  thumbnailUrl?: string;
}): Promise<ApiResponse<TutorialEmailTemplate>> {
  return apiRequest(api.tutorialEmail.templates.list(), { method: 'POST', body: data });
}

export async function updateTutorialEmailTemplate(id: string, data: {
  name?: string;
  htmlContent?: string;
  variables?: string[];
  assetIds?: string[];
  thumbnailUrl?: string;
}): Promise<ApiResponse<TutorialEmailTemplate>> {
  return apiRequest(api.tutorialEmail.templates.byId(id), { method: 'PUT', body: data });
}

export async function deleteTutorialEmailTemplate(id: string): Promise<ApiResponse<{ deleted: boolean }>> {
  return apiRequest(api.tutorialEmail.templates.byId(id), { method: 'DELETE' });
}

// ========== Assets ==========

export async function listTutorialEmailAssets(tag?: string): Promise<ApiResponse<{ items: TutorialEmailAsset[] }>> {
  const qs = tag ? `?tag=${encodeURIComponent(tag)}` : '';
  return apiRequest(`${api.tutorialEmail.assets.list()}${qs}`, { method: 'GET' });
}

export async function createTutorialEmailAsset(data: {
  fileName: string;
  fileUrl: string;
  tags?: string[];
  fileSize?: number;
  contentType?: string;
}): Promise<ApiResponse<TutorialEmailAsset>> {
  return apiRequest(api.tutorialEmail.assets.list(), { method: 'POST', body: data });
}

export async function deleteTutorialEmailAsset(id: string): Promise<ApiResponse<{ deleted: boolean }>> {
  return apiRequest(api.tutorialEmail.assets.byId(id), { method: 'DELETE' });
}

// ========== Enrollments ==========

export async function listTutorialEmailEnrollments(args?: {
  sequenceKey?: string;
  status?: string;
}): Promise<ApiResponse<{ items: TutorialEmailEnrollment[] }>> {
  const params = new URLSearchParams();
  if (args?.sequenceKey) params.set('sequenceKey', args.sequenceKey);
  if (args?.status) params.set('status', args.status);
  const qs = params.toString() ? `?${params.toString()}` : '';
  return apiRequest(`${api.tutorialEmail.enrollments.list()}${qs}`, { method: 'GET' });
}

export async function enrollTutorialEmailUser(data: {
  userId: string;
  email: string;
  sequenceKey: string;
}): Promise<ApiResponse<TutorialEmailEnrollment>> {
  return apiRequest(api.tutorialEmail.enrollments.list(), { method: 'POST', body: data });
}

export async function unsubscribeTutorialEmailEnrollment(id: string): Promise<ApiResponse<{ unsubscribed: boolean }>> {
  return apiRequest(api.tutorialEmail.enrollments.unsubscribe(id), { method: 'POST', body: {} });
}

export async function batchEnrollTutorialEmail(data: {
  sequenceKey: string;
}): Promise<ApiResponse<{ enrolled: number; skipped: number; total: number }>> {
  return apiRequest(api.tutorialEmail.enrollments.batch(), { method: 'POST', body: data });
}

// ========== Test Send ==========

export async function testSendTutorialEmail(data: {
  email: string;
  name?: string;
  subject?: string;
  templateId: string;
}): Promise<ApiResponse<{ success: boolean }>> {
  return apiRequest(api.tutorialEmail.testSend(), { method: 'POST', body: data });
}

// ========== AI Generate ==========

export async function generateTutorialEmailTemplate(data: {
  topic?: string;
  style?: string;
  language?: string;
  extraRequirements?: string;
  messages?: Array<{ role: string; content: string }>;
}): Promise<ApiResponse<{ htmlContent: string; model?: string; tokens?: number }>> {
  return apiRequest(api.tutorialEmail.generate(), { method: 'POST', body: data });
}

// ========== Quick Send ==========

export async function quickSendTutorialEmail(data: {
  email: string;
  recipientName?: string;
  subject?: string;
  htmlContent: string;
  saveAsTemplate?: boolean;
  templateName?: string;
}): Promise<ApiResponse<{ sent: boolean; templateId?: string; templateName?: string }>> {
  return apiRequest(api.tutorialEmail.quickSend(), { method: 'POST', body: data });
}
